### PR TITLE
Allow a set of special characters to be used in path

### PR DESCRIPTION
**Pull request by WalshyDev**
This allows the user to define a set of special characters which can be used in the path. These special characters cannot conflict with the pathSeparator and setting that will throw an exception if it does. 
This should fix quite a few possible issues that can arise from the path separator and the pretty much anything goes JSON path.

I also changed the regex to a better suiting one and added the escaped allowed chars (if applicable).
This passes all tests still!

### CODE REVIEW

File: src/main/java/io/github/binaryoverload/JSONConfig.java [hunk 1]
```diff
@@ -202,20 +212,30 @@ public JSONConfig(JsonObject object) {
...
         this.object = object;
         setPathSeparator(pathSeparator);
         this.mode = -1;
     }
 
+    private JSONConfig(char[] allowedSpecialCharacters) {
+        this.allowedSpecialCharacters = allowedSpecialCharacters;
+    }
+
+    private JSONConfig() {
```
Comment (binaryoverload) [inline src/main/java/io/github/binaryoverload/JSONConfig.java]: Why...?

File: src/test/java/io/github/binaryoverload/GeneralUtilsTest.java [hunk 2]
```diff
@@ -52,16 +52,16 @@ public void checkStringLengthNegative() {
...
+        GeneralUtils.verifyPath("cat_cat.cat", '.', new char[]{'_'});
+        GeneralUtils.verifyPath("cat-cat-cat", '-', new char[]{});
     }
 
     @Test(expected = IllegalArgumentException.class)
     public void verifyPathNegative() {
-        GeneralUtils.verifyPath("cat.cat.cat..", '.');
-        GeneralUtils.verifyPath("cat-cat-cat-", '-');
+        GeneralUtils.verifyPath("cat.cat.cat..", '.', new char[]{});
+        GeneralUtils.verifyPath("cat-cat-cat-", '-', new char[]{});
```
Comment (binaryoverload) [inline src/test/java/io/github/binaryoverload/GeneralUtilsTest.java]: Possibly add another test to test for the specific case of unallowed characters?

File: src/main/java/io/github/binaryoverload/JSONConfig.java (hunk 1 above)
Comment (WalshyDev) [reply src/main/java/io/github/binaryoverload/JSONConfig.java]: For the purpose of not needing to always specify special chars. I could just not pass `this()` but it doesn't cost anything extra and still allows a bit more expansion for anything else that needs to be done in all constructors.

File: src/test/java/io/github/binaryoverload/GeneralUtilsTest.java (hunk 2 above)
Comment (WalshyDev) [reply src/test/java/io/github/binaryoverload/GeneralUtilsTest.java]: Sure.

**Comment by binaryoverload** [comment]
Also, fix your regex

File: src/main/java/io/github/binaryoverload/JSONConfig.java (hunk 1 above)
Comment (binaryoverload) [reply src/main/java/io/github/binaryoverload/JSONConfig.java]: :+1:

**Pull request merged**